tests/BaseElementTest.php

```diff
@@ -236,4 +252,141 @@ public function testOnBeforeWriteNoParent()
+            'element4' => [
+                ElementContent::class,
+                'content1',
+                'http://localhost/admin/pages/edit/show/1',
+            ],
+            // Element in DataObject with $directLink === true
+            'element5' => [
+                ElementContent::class,
+                'content1',
+                'admin/pages/edit/EditForm/1/field/ElementalArea/item/1/edit',
+                true
+            ],
+            // Element without any Page or DataObject
+            'element6' => [
+                ElementContent::class,
+                'contentDataObject2',
+                null
+            ],
+            // DataObject without CMSEditLink method
+            'element7' => [
+                TestElement::class,
+                'elementDataObject2',
+                null
+            ],
+        ];
+    }
+
+    /**
+     * @dataProvider getElementCMSLinkDataProvider
+     */
+    public function testCMSEditLink(string $class, string $element, ?string $link, bool $directLink = false)
+    {
+        $object = $this->objFromFixture($class, $element);
+        $editLink = $object->CMSEditLink($directLink);
+
+        if ($link) {
+            $this->assertStringContainsString($link, $editLink);
+        } else {
+            $this->assertNull($link);
+        }
+    }
+
+    public function canDeleteProvider()
+    {
+        return [
+            // Element on Page
+            'element1' => [
+                ElementContent::class,
+                'content1',
+                'ADMIN',
+            ],
+            // Element in DataObject
+            'element2' => [
+                TestElement::class,
+                'elementDataObject1',
+                'CMS_ACCESS_CMSMain',
+            ],
+        ];
+    }
+
+    /**
+     * @dataProvider canDeleteProvider
+     */
+    public function testCanDelete(
+        string $class,
+        string $element,
+        string $permission
+    ) {
+        $this->logOut();
+        $this->logInWithPermission($permission);
+        $object = $this->objFromFixture($class, $element);
+        $canDelete = $object->canDelete();
+        $this->assertNotNull($canDelete);
+    }
+
+    public function linksProvider()
+    {
+        return [
+            // Element on Page
+            'element1' => [
+                ElementContent::class,
+                'content1',
+                '/test-elemental/#e1',
+            ],
+            // Element in DataObject
+            'element2' => [
+                TestElement::class,
+                'elementDataObject1',
+                null,
+            ],
+        ];
+    }
+
+    /**
+     * @dataProvider linksProvider
+     */
+    public function testLinkWithDataObject(string $class, string $element, ?string $link)
+    {
+        $object = $this->objFromFixture($class, $element);
+        $canDelete = $object->Link();
+        $this->assertEquals($link, $canDelete);
```

Comment:
```suggestion
        $this->assertEquals($link, $object->Link());
```
I assume the `$canDelete` variable name was a copy-paste error - but in any case there's no need to assign an intermediate variable here.